0006-ffmul-revised.md

```diff
@@ -0,0 +1,2249 @@
+
+$$
+q \cong (q_0, q_1, q_2) = (2^\ell - |q|_0, 2^\ell - 1 - |q|_1, -|q|_2 - 1)
+$$
+
+(Detail: If $|q|_0 = 0$ this doesn't work but then we just don't do the borrowing.)
+
+In words, by borrowing from higher limbs we only need the highest limb $q_2 = -|q|_2 - 1$ to be negative. Now, $-|q|_2 - 1$ is indistinguishable in our constraints from $n-|q|_2 - 1$, and $q_2 = n-|q|_2 - 1$ is an allowed value for all $|q|_2 \in [0, 2^\ell - f_2)$. This means we can "represent" most negative $q \in (-2^{3\ell}, 0)$ by
+
+$$
+q_\text{fake} \cong (2^\ell - |q|_0, 2^\ell - |q|_1 + 1, n-|q|_2 - 1).
+$$
+
+Note that
+
+$$
+q_\text{fake} = q + 2^{2\ell} n
+$$
+
+> Attack on `ForeignFieldMul`
+
+If we are able to use large negative $q$, the estimates for the final bound $(L.9)$ are no longer valid, which we can exploit. Given any $a$, $b$, we can find $q$, $r$ such that
+
+$$\tag{L.9b}
+ab = qf + r + 2^{3\ell} n
+$$
+
+This is always possible: we pick $q$ and $r$ to be the quotient and remainder of the modified divsion $(ab - 2^{3\ell} n) / f$. Concretely, $q = \lfloor\frac{ab - 2^{3\ell} n}{f}\rfloor$, which is a negative number (but big enough that we can always represent it by $q_\text{fake}$, if e.g. $2n < f$).
+
+If we replace $q$ by $q_\text{fake} = q + 2^{2\ell}n$, we get a modified equation which still features the same $a$, $b$, $r$:
+
+$$\tag{L.9c}
+ab = (q_\text{fake} - 2^{2\ell} n)f + r + 2^{3\ell} n
+$$
+
+So, we want to fool the verifier into accepting $a$, $b$ and $r$ satisfying $(L.9c)$. $r$ is not the correct remainder of $ab / f$, because $2^{3\ell} n$ is not a multiple of $f$.
+
+Since $q_\text{fake} = q \bmod n$, equation $\text{(L.2a: native)}$ is satisfied, and since we use the correct, $\ell$-bit values for $q_0$ and $q_1$, equations $\text{(L.3: split)}$ and $\text{(L.4: bottom limbs)}$ are as well (added multiples of $n$ like $2^{3\ell} n$ affect none of our constraints).
+
+The only way our attack can sometimes fail is because $q_2$ is now "negative", the expression $p_2$ is no longer guaranteed to be positive and we might underflow 0 in $\text{(L.5: high limb)}$, which the range-check on the RHS doesn't allow:
+
+$$\tag{L.5}
+(ab)_2 + q_0f'_2 + q_1f'_1 + (- |q|_2 - 1)f'_0 + p_{11} - r_2 + v_0 = 2^{\ell} v_{1} \mod n 
+$$
+
+However, underflow caused by one negative value $(- |q|_2 - 1)f'_0$ against 5 positive values of similar size is the exception. Furthermore, there are important primes where $f'_0$ is an unusually small number:
+* For Curve25519 we have $f'_0 = 19$
+* For seqp256k1 we have $f'_0 = 2^{32} + 2^9 + 2^8 + 2^7 + 2^6 + 2^4 + 1$
+
+In these cases, the negative $-|q|_2f'_0$ is small and should never cause underflow for "random" input values. In other words, the prover can use an incorrect output $r$ in almost every foreign field multiplication.
+
+**Conclusion**: The full range check $q_2 \in [0,2^\ell)$ is necessary.
+
+> Counter-example 2: Compact range-check on $q_0 + 2^\ell q_1$
+
+TBC
```

Comment:
Our goal is to have a FFmul RFC that will lock its design. For that reason, shouldn't we confirm the TBC before we "lock it" and move on to implementation?

Reply:
Yeah I think we can remove the "counter-example 2" section, I feel it's not necessary to put in the work of constructing another counter-example

Reply:
Sounds good, thank you all!